actionpack/lib/action_dispatch/routing/url_for.rb

```diff
@@ -171,6 +171,10 @@ def url_for(options = nil)
           route_name = options.delete :use_route
           _routes.url_for(options.symbolize_keys.reverse_merge!(url_options),
                          route_name)
+        when ActionController::Parameters
```

Comment:
In fact, I think we should not allow parameters instance directly on `url_for`. See https://github.com/rails/rails/pull/20797. I'll work with @byroot to finish that so we can remove all this code.